CMakeLists.txt

```diff
@@ -24,6 +24,7 @@ the library is not compromised. Turn this OFF to ignore warnings." ON)
 option(S2N_INTERN_LIBCRYPTO "This ensures that s2n-tls is compiled and deployed with a specific
 version of libcrypto by interning the code and hiding symbols. This also enables s2n-tls to be
 loaded in an application with an otherwise conflicting libcrypto version." OFF)
+option(S2N_LTO, "Enables link time optimizations when building s2n-tls." OFF)
 # Turn BUILD_TESTING=ON by default
```

Comment:
What are the downsides of turning this on by default?

Reply:
The only thing I can think of right now is if it's not supported by the target compiler/linker, although I'm not sure how likely that is.

Reply:
I'd love us to make it the default. Maybe we can do some more research and wait for the next minor version bump.

Reply:
https://wiki.debian.org/LTO#Downsides_of_LTO_are

* increased compile time
* bit of extra effort to get the parameters right
* unclear compatibility with other compilers than gcc/g++
* it is not so very wide-spread, you may run into compiler errors (or not notice them, which would be worse). There is a thread on Debian Devel collecting experiences (https://lists.debian.org/debian-devel/2016/03/msg00399.html) and opinions that yet speak against its wider utilisation in Debian. Particularly when doing tricky bits as in the Linux Kernel (https://lwn.net/Articles/512548/) you should be careful. Well, the maintainers decide for themselves and we have the experimental section when in doubt. It is all routine for the GCC developers now (https://gcc.gnu.org/wiki/LinkTimeOptimization).
* it may not always work (e.g. http://www.phoronix.com/scan.php?page=article&item=gcc_471_lto&num=2) but this is 5 years ago and the sizes of binaries are not mentioned as a check that LTO had kicked in. Find problems with how parameters are passed to the linker if binaries do not get smaller. One rarely gets it right at the first attempt. There is little to no reason why more freedom to optimize should get something worse.

Reply:
One call-out is the difference between gnu make and cmake.  Since we're running integration tests under make, we might not be getting the coverage we expect. Note: not suggesting LTO gets added to Make, but we really should pick a side.

Reply:
Also does this need a try-compile?

Reply:
> we really should pick a side.

We absolutely should. I guess I'm picking a side by only adding it to CMake 😄.

> Also does this need a try-compile?

If we want to enable it by default we would, yes. But since it's currently an opt-in thing, we assume the caller knows if it's supported or not.

Reply:
> What are the downsides of turning this on by default?

Inability to debug code which is linked in this way pre GCC 8 as it has the side effect of stripping debugging symbols (the symbols are unusable, so may as well be stripped).